File: legacy_backend/aggregators/base/process.py
"""
Content processing and formatting for aggregators.

This module provides content processing functionality including:
- Removing unwanted HTML elements
- Sanitizing HTML
- Standardizing content format (adding header images, source links)
"""

import base64
import logging
from typing import TYPE_CHECKING
from urllib.parse import urljoin

from bs4 import BeautifulSoup

if TYPE_CHECKING:
    from .models import RawArticle

from .fetch import extract_image_from_url
from .utils import (
    compress_image,
    extract_youtube_video_id,
    remove_elements_by_selectors,
    sanitize_html,
)

logger = logging.getLogger(__name__)


# ============================================================================
# Content Standardization
# ============================================================================


def standardize_content_format(
    content: str,
    article: "RawArticle",
    base_url: str | None = None,
    generate_title_image: bool = True,
    add_source_footer: bool = True,
    header_image_url: str | None = None,
) -> str:
    """
    Standardize content format across all feeds.

    This function:
    1. Finds the first URL (link or image) in the content (if generate_title_image=True)
    2. Extracts an image from that URL (or uses meta tags, first image, or favicon)
    3. Compresses and inlines the image as base64
    4. Places the image at the top of the content
    5. Removes the original image tag if it was in the content
    6. Adds the content below the image
    7. Adds a source link at the bottom (float right) (if add_source_footer=True)

    Args:
        content: The HTML content to standardize
        article: The article being processed
        base_url: Optional base URL for resolving relative URLs (defaults to article.url)
        generate_title_image: Whether to extract and add a header image (default: True)
        add_source_footer: Whether to add a source link at the bottom (default: True)
        header_image_url: Optional pre-determined header image URL (overrides automatic detection)

    Returns:
        Standardized HTML content
    """
    if not base_url:
        base_url = article.url

    logger.debug(f"Standardizing content format for: {article.url}")

    try:
        soup = BeautifulSoup(content, "html.parser")

        # Build the standardized content
        content_parts = []

        # Extract and add header image if enabled
        if generate_title_image:
            # First, check if article.url is a YouTube video - embed it instead of extracting image
            article_video_id = extract_youtube_video_id(article.url)
            if article_video_id:
                from api.youtube import get_youtube_proxy_url

                embed_url = get_youtube_proxy_url(article_video_id)
                content_parts.append(
                    '<div class="youtube-embed-container">'
                    f'<iframe width="560" height="315" src="{embed_url}" '
                    'title="YouTube video player" '
                    'frameborder="0" '
                    'allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" '
                    "allowfullscreen></iframe>"
                    "</div>"
                )
                logger.debug(f"Added YouTube embed for video: {article_video_id}")
                # Skip image extraction for YouTube videos
            else:
                # Find the first URL (link or image)
                first_url = None
                first_element = None

                # First, check for YouTube links in content (they take priority over header images)
                first_link = soup.find("a", href=True)
                if first_link:
                    link_url = urljoin(base_url, first_link["href"])
                    if extract_youtube_video_id(link_url):
                        first_url = link_url
                        first_element = first_link
                        logger.debug(f"Found YouTube link in content: {first_url}")

                # If no YouTube link found, use pre-determined header image URL if provided
                is_using_header_image = False
                if not first_url and header_image_url:
                    # Resolve relative URLs to absolute URLs
                    first_url = urljoin(base_url, header_image_url)
                    is_using_header_image = True
                    logger.debug(f"Using pre-determined header image: {first_url}")
                elif not first_url:
                    # First, try to find an image
                    first_img = soup.find("img")
                    if first_img:
                        img_src = (
                            first_img.get("src")
                            or first_img.get("data-src")
                            or first_img.get("data-lazy-src")
                        )
                        if img_src:
                            first_url = urljoin(base_url, img_src)
                            first_element = first_img
                            logger.debug(f"Found first image: {first_url}")

                    # If no image, try to find first link
                    if not first_url and first_link:
                        first_url = urljoin(base_url, first_link["href"])
                        first_element = first_link
                        logger.debug(f"Found first link: {first_url}")

                    # If still no URL, use the article URL itself
                    if not first_url:
                        first_url = article.url
                        logger.debug(
                            f"No URL found in content, using article URL: {article.url}"
                        )

                # Check if URL is a YouTube video - embed it instead of extracting image
                video_id = extract_youtube_video_id(first_url)
                if video_id:
                    from api.youtube import get_youtube_proxy_url

                    embed_url = get_youtube_proxy_url(video_id)
                    content_parts.append(
                        '<div class="youtube-embed-container">'
                        f'<iframe width="560" height="315" src="{embed_url}" '
                        'title="YouTube video player" '
                        'frameborder="0" '
                        'allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" '
                        "allowfullscreen></iframe>"
                        "</div>"
                    )
                    logger.debug(f"Added YouTube embed for video: {video_id}")

                    # Remove the original link/image from content
                    if first_element:
                        parent = first_element.parent
                        first_element.decompose()
                        # Remove empty parent containers recursively
                        while parent and parent.name not in (
                            "body",
                            "[document]",
                            "html",
                        ):
                            if not parent.get_text(strip=True) and not parent.find():
                                next_parent = parent.parent
                                parent.decompose()
                                parent = next_parent
                            else:
                                break
                        logger.debug("Removed original YouTube link/image from content")
                else:
                    # Extract image from the URL or data URI
                    image_result = None

                    # Check if the URL is already a data URI (base64 encoded)
                    if first_url.startswith("data:"):
                        logger.debug(
                            "First image is already a data URI, extracting data"
                        )
                        try:
                            # Parse data URI: data:image/png;base64,iVBORw0KG...
                            if ";base64," in first_url:
                                header, encoded = first_url.split(";base64,", 1)
                                content_type = (
                                    header.split(":", 1)[1]
                                    if ":" in header
                                    else "image/jpeg"
                                )

                                # CRITICAL: Validate that data URI is actually an image
                                if not content_type.startswith("image/"):
                                    logger.warning(
                                        f"Data URI has non-image content type: {content_type}, skipping"
                                    )
                                else:
                                    image_data = base64.b64decode(encoded)

                                    # Additional validation: Try to parse as image with PIL
                                    try:
                                        import io

                                        from PIL import Image

                                        img = Image.open(io.BytesIO(image_data))
                                        img.verify()  # Verify it's actually a valid image
                                        image_result = (image_data, content_type)
                                        logger.debug(
                                            f"Extracted valid data URI image: {content_type}, {len(image_data)} bytes"
                                        )
                                    except Exception as pil_error:
                                        logger.warning(
                                            f"Data URI claims to be {content_type} but failed PIL validation: {pil_error}"
                                        )
                        except Exception as e:
                            logger.error(f"Failed to parse data URI: {e}")
                    else:
                        # Extract image from regular URL
                        # If this is a header_image_url, pass flag to skip width/height filtering
                        image_result = extract_image_from_url(
                            first_url, is_header_image=is_using_header_image
                        )
                        # Save thumbnail URL for non-data URIs (actual URLs)
                        if image_result:
                            article.thumbnail_url = first_url
                            logger.debug(f"Saved thumbnail URL: {first_url}")

                    # Add the header image if we found one
                    if image_result:
                        image_data, content_type = image_result

                        # Compress the image
                        compressed_data, output_type = compress_image(
                            image_data, content_type
                        )

                        # Convert to base64
                        image_b64 = base64.b64encode(compressed_data).decode("utf-8")
                        data_uri = f"data:{output_type};base64,{image_b64}"

                        # Add image at the top
                        content_parts.append(
                            f'<p><img src="{data_uri}" alt="Article image" style="max-width: 100%; height: auto;"></p>'
                        )
                        logger.debug("Added header image to content")

                        # Remove the original image from content if it was an img tag
                        if first_element and first_element.name == "img":
                            parent = first_element.parent
                            first_element.decompose()
                            # Remove empty parent containers recursively
                            while parent and parent.name not in (
                                "body",
                                "[document]",
                                "html",
                            ):
                                if (
                                    not parent.get_text(strip=True)
                                    and not parent.find()
                                ):
                                    next_parent = parent.parent
                                    parent.decompose()
                                    parent = next_parent
                                else:
                                    break
                            logger.debug("Removed original image from content")

        # Add the remaining content
        content_parts.append(str(soup))

        # Add source link at the bottom (float right) if enabled
        if add_source_footer:
            content_parts.append(
                f'<a href="{article.url}" style="float: right;">Source</a>'
            )

        return "".join(content_parts)

    except Exception as e:
        logger.error(f"Error standardizing content format: {e}", exc_info=True)
        # Fallback: add source link if enabled
        if add_source_footer:
            return f'{content}<a href="{article.url}" style="float: right;">Source</a>'
        return content


# ============================================================================
# Processing Mixin for BaseAggregator
# ============================================================================


class ProcessingMixin:
    """
    Mixin providing content processing functionality for aggregators.

    This mixin provides methods for processing HTML content:
    - Removing unwanted elements (ads, social buttons, etc.)
    - Sanitizing HTML (removing scripts, iframes, etc.)
    - Standardizing format (adding header images, source links)
    """

    def get_header_image_url(self, article: "RawArticle") -> str | None:
        """
        Extract the header image URL from the article's full HTML.

        Override this method to provide custom header image extraction logic.
        This is called before content extraction, so article.html contains the
        full page HTML.

        The returned URL will be used as the header image in the final article,
        bypassing the automatic image detection in standardize_content_format.

        Args:
            article: The article being processed (with full HTML)

        Returns:
            The header image URL, or None to use automatic detection

        Example:
            >>> def get_header_image_url(self, article: RawArticle) -> str | None:
            ...     soup = BeautifulSoup(article.html, 'html.parser')
            ...     header_img = soup.find('div', id='header').find('img')
            ...     return header_img['src'] if header_img else None
        """
        return None  # Default: use automatic detection

    def remove_unwanted_elements(self, article: "RawArticle") -> None:
        """
        Remove unwanted HTML elements (ads, social buttons, etc.)

        Override this if you need custom removal logic beyond CSS selectors.

        Reads from article.html and updates it with cleaned content.

        Args:
            article: The article being processed (reads/writes article.html)
        """
        if not self.selectors_to_remove:
            return

        article.html = remove_elements_by_selectors(
            article.html,
            self.selectors_to_remove,
            remove_empty=True,
        )

    def sanitize_content(self, article: "RawArticle") -> None:
        """
        Sanitize HTML content (remove scripts, iframes, etc.)

        Override this if you need custom sanitization logic.

        Reads from article.html and updates it with sanitized content.

        Args:
            article: The article being processed (reads/writes article.html)
        """
        article.html = sanitize_html(article.html)

    def standardize_format(
        self, article: "RawArticle", header_image_url: str | None = None
    ) -> None:
        """
        Standardize content format (add header image, source link, etc.)

        Override this to customize the final formatting.

        Reads from article.html and updates it with standardized content.

        Args:
            article: The article being processed (reads/writes article.html)
            header_image_url: Optional pre-determined header image URL
        """
        article.html = standardize_content_format(
            article.html,
            article,
            generate_title_image=self.feed.generate_title_image if self.feed else True,
            add_source_footer=self.feed.add_source_footer if self.feed else True,
            header_image_url=header_image_url,
        )

File: src/server/aggregators/base/process.ts
/**
 * Content processing utilities.
 */

import * as cheerio from "cheerio";
import type { Element } from "domhandler";
import sharp from "sharp";
import type { RawArticle } from "./types";
import {
  extractImageFromUrl,
  compressImage,
  MAX_HEADER_IMAGE_WIDTH,
  MAX_HEADER_IMAGE_HEIGHT,
  extractYouTubeVideoId,
  getYouTubeProxyUrl,
  createHeaderElementFromUrl,
} from "./utils";
import { logger } from "@server/utils/logger";

/**
 * Standardize content format across all feeds.
 *
 * This function wraps content in semantic HTML structure:
 * <article>
 *   <header>...</header>
 *   <section>...main content...</section>
 *   <section>...comments...</section>
 *   <footer>...</footer>
 * </article>
 *
 * Steps:
 * 1. Checks if content is already wrapped in <article> tag
 * 2. Checks if <header> already exists
 * 3. If no header exists and generate_title_image=true:
 *    - Finds the first URL (link or image) in the content
 *    - Extracts an image from that URL (or uses meta tags, first image, or favicon)
 *    - Compresses and inlines the image as base64
 *    - Places the image in <header> tag
 *    - Removes the original image tag if it was in the content
 * 4. Wraps the remaining content in <section> tag (unless already in section)
 * 5. Checks if <footer> already exists
 * 6. If no footer exists and add_source_footer=true:
 *    - Adds a source link in <footer> tag
 *
 * This prevents duplicate headers and footers when aggregators add them manually.
 */
export async function standardizeContentFormat(
  content: string,
  article: RawArticle,
  baseUrl?: string,
  generateTitleImage: boolean = true,
  addSourceFooter: boolean = true,
  headerImageUrl?: string,
): Promise<string> {
  if (!baseUrl) {
    baseUrl = article.url;
  }

  logger.debug({ url: article.url }, "Standardizing content format");

  try {
    const $ = cheerio.load(content);

    // Check if content is already wrapped in <article> tag
    const hasArticleWrapper = $("article").length > 0;

    // Extract content from article if it exists, otherwise use full content
    // If article exists, get its inner HTML; otherwise get body HTML
    let bodyContent: string;
    if (hasArticleWrapper) {
      bodyContent = $("article").html() || "";
    } else {
      // Get body content, or if no body, get root HTML
      const body = $("body");
      bodyContent = body.length > 0 ? body.html() || "" : $.html();
    }

    // Re-parse the body content to work with it
    const $body = cheerio.load(bodyContent);

    // Check if header already exists
    const hasExistingHeader = $body("header").length > 0;

    const contentParts: string[] = [];

    // Extract and add header image if enabled and not already present
    if (generateTitleImage && !hasExistingHeader) {
      // Find the first URL (link or image) from body content
      let firstUrl: string | null = null;
      let firstElement: cheerio.Cheerio<Element> | null = null;

      // Use pre-determined header image URL if provided (aggregator's best guess)
      // Otherwise, find first URL from content
      let isUsingHeaderImage = false;
      if (headerImageUrl) {
        // Resolve relative URLs to absolute URLs
        firstUrl = new URL(headerImageUrl, baseUrl).toString();
        isUsingHeaderImage = true;
        logger.debug(
          { url: firstUrl },
          "Using pre-determined header image URL",
        );

        // Find matching element in content to remove it later
        // This prevents duplicate images (one in header, one in content)
        // Helper function to normalize URLs for comparison (remove trailing slashes, fragments, etc.)
        const normalizeUrl = (url: string) =>
          url.replace(/\/$/, "").split("#")[0].split("?")[0];

        if (!firstElement) {
          // Try to find matching image element
          $body("img").each((_, element) => {
            if (firstElement) return; // Already found one
            const $img = $body(element);
            const imgSrc =
              $img.attr("src") ||
              $img.attr("data-src") ||
              $img.attr("data-lazy-src");
            if (imgSrc) {
              try {
                const resolvedImgSrc = new URL(imgSrc, baseUrl).toString();
                // Compare normalized URLs
                if (normalizeUrl(resolvedImgSrc) === normalizeUrl(firstUrl)) {
                  firstElement = $body(element);
                  logger.debug(
                    { url: resolvedImgSrc },
                    "Found matching image element in content",
                  );
                }
              } catch (error) {
                // Invalid URL, skip
              }
            }
          });

          // If no matching image, try to find matching link
          if (!firstElement) {
            $body("a[href]").each((_, element) => {
              if (firstElement) return; // Already found one
              const $link = $body(element);
              const linkHref = $link.attr("href");
              if (linkHref) {
                // Skip invalid URLs
                if (
                  linkHref.includes("${") ||
                  linkHref.startsWith("javascript:") ||
                  linkHref.startsWith("data:") ||
                  linkHref.trim() === ""
                ) {
                  return;
                }
                try {
                  const resolvedLinkHref = new URL(
                    linkHref,
                    baseUrl,
                  ).toString();
                  // Compare normalized URLs
                  if (
                    normalizeUrl(resolvedLinkHref) === normalizeUrl(firstUrl)
                  ) {
                    firstElement = $body(element);
                    logger.debug(
                      { url: resolvedLinkHref },
                      "Found matching link element in content",
                    );
                  }
                } catch (error) {
                  // Invalid URL, skip
                }
              }
            });
          }
        }
      } else if (!firstUrl) {
        // First, try to find an image
        const firstImg = $body("img").first();
        if (firstImg.length > 0) {
          const imgSrc =
            firstImg.attr("src") ||
            firstImg.attr("data-src") ||
            firstImg.attr("data-lazy-src");
          if (imgSrc) {
            firstUrl = new URL(imgSrc, baseUrl).toString();
            firstElement = firstImg;
            logger.debug({ url: firstUrl }, "Found first image");
          }
        }

        // If no image, try to find first link
        const firstLink = $body("a[href]").first();
        if (!firstUrl && firstLink.length > 0) {
          const linkHref = firstLink.attr("href");
          if (linkHref) {
            // Skip invalid URLs (template literals, JavaScript code, etc.)
            if (
              linkHref.includes("${") ||
              linkHref.startsWith("javascript:") ||
              linkHref.startsWith("data:") ||
              linkHref.trim() === ""
            ) {
              logger.debug({ linkHref }, "Skipping invalid link URL");
            } else {
              try {
                firstUrl = new URL(linkHref, baseUrl).toString();
                // Additional validation: ensure URL is actually valid (not containing template syntax)
                if (!firstUrl.includes("${") && !firstUrl.includes("%7B")) {
                  firstElement = firstLink;
                  logger.debug({ url: firstUrl }, "Found first link");
                } else {
                  logger.debug(
                    { firstUrl },
                    "Skipping URL with template syntax",
                  );
                  firstUrl = null;
                }
              } catch (error) {
                logger.debug({ error, linkHref }, "Failed to parse link URL");
              }
            }
          }
        }

        // If still no URL, use the article URL itself
        if (!firstUrl) {
          firstUrl = article.url;
          logger.debug(
            { url: firstUrl },
            "No URL found in content, using article URL",
          );
        }
      }

      // Final validation: ensure firstUrl doesn't contain template syntax (including URL-encoded)
      if (
        firstUrl &&
        (firstUrl.includes("${") ||
          firstUrl.includes("%7B") ||
          firstUrl.includes("%24%7B"))
      ) {
        logger.debug(
          { firstUrl },
          "Skipping URL with template syntax, using article URL instead",
        );
        firstUrl = article.url;
      }

      // Use the unified header element creation function
      // This handles YouTube embeds, image extraction, compression, and base64 encoding
      if (firstUrl) {
        // Final safety check: ensure URL is valid before attempting extraction
        if (
          !firstUrl.includes("${") &&
          !firstUrl.includes("%7B") &&
          !firstUrl.includes("%24%7B") &&
          (firstUrl.startsWith("http") || firstUrl.startsWith("data:"))
        ) {
          // Handle data URI separately (already base64 encoded)
          if (firstUrl.startsWith("data:")) {
            logger.debug("First image is already a data URI, extracting data");
            try {
              // Parse data URI: data:image/png;base64,iVBORw0KG...
              if (firstUrl.includes(";base64,")) {
                const [header, encoded] = firstUrl.split(";base64,", 2);
                let contentType = header.split(":")[1] || "image/jpeg";

                // CRITICAL: Validate that data URI is actually an image
                if (!contentType.startsWith("image/")) {
                  logger.warn(
                    { contentType },
                    "Data URI has non-image content type, skipping",
                  );
                } else {
                  const imageData = Buffer.from(encoded, "base64");

                  // Additional validation: Try to parse as image with sharp
                  try {
                    await sharp(imageData).metadata(); // This will throw if not a valid image

                    // Compress the image with higher resolution for header images
                    const compressed = await compressImage(
                      imageData,
                      contentType,
                      MAX_HEADER_IMAGE_WIDTH,
                      MAX_HEADER_IMAGE_HEIGHT,
                    );
                    const compressedData = compressed.imageData;
                    const outputType = compressed.contentType;

                    // Convert to base64
                    const imageB64 = compressedData.toString("base64");
                    const dataUri = `data:${outputType};base64,${imageB64}`;

                    // Add image in header
                    contentParts.push(
                      `<header><p><img src="${dataUri}" alt="Article image" style="max-width: 100%; height: auto;"></p></header>`,
                    );
                    logger.debug("Added compressed data URI image to content");
                  } catch (error) {
                    logger.warn(
                      { error, contentType },
                      "Data URI claims to be image but failed validation",
                    );
                  }
                }
              }
            } catch (error) {
              logger.error({ error }, "Failed to parse data URI");
            }
          } else {
            // Use unified function for regular URLs (handles YouTube, Twitter, Reddit, images, etc.)
            const headerElement = await createHeaderElementFromUrl(
              firstUrl,
              "Article image",
            );

            if (headerElement) {
              // Wrap header element in <header> tag if not already wrapped
              let wrappedHeader: string;
              if (
                headerElement.includes("<header>") ||
                headerElement.includes("<header ")
              ) {
                wrappedHeader = headerElement;
              } else {
                // Remove any existing wrapper divs with data-article-header
                const cleaned = headerElement
                  .replace(/<div[^>]*data-article-header[^>]*>/gi, "")
                  .replace(/<\/div>/gi, "");
                wrappedHeader = `<header>${cleaned}</header>`;
              }
              contentParts.push(wrappedHeader);

              // Save thumbnail URL (will be converted to base64 by aggregation service)
              article.thumbnailUrl = firstUrl;
              logger.debug({ url: firstUrl }, "Saved thumbnail URL");

              // Check if it's a YouTube embed to handle duplicate removal
              const videoId = extractYouTubeVideoId(firstUrl);
              if (videoId && isUsingHeaderImage) {
                // Remove duplicate YouTube links from content
                $body("a[href]").each((_, element) => {
                  const href = $(element).attr("href");
                  if (href) {
                    try {
                      const linkVideoId = extractYouTubeVideoId(
                        new URL(href, baseUrl).toString(),
                      );
                      if (linkVideoId === videoId) {
                        const parent = $(element).parent();
                        $(element).remove();
                        // Remove empty parent containers recursively
                        let currentParent = parent;
                        while (currentParent.length > 0) {
                          const tagName = currentParent
                            .get(0)
                            ?.tagName?.toLowerCase();
                          if (tagName === "body" || tagName === "html") {
                            break;
                          }
                          const text = currentParent.text().trim();
                          const hasChildren =
                            currentParent.children().length > 0;
                          if (!text && !hasChildren) {
                            const nextParent = currentParent.parent();
                            currentParent.remove();
                            currentParent = nextParent;
                          } else {
                            break;
                          }
                        }
                        logger.debug(
                          "Removed duplicate YouTube link from content",
                        );
                      }
                    } catch (error) {
                      // Invalid URL, skip
                    }
                  }
                });

                // Remove first image from content if it exists (since we're using YouTube video instead)
                const firstImg = $body("img").first();
                if (firstImg.length > 0) {
                  const parent = firstImg.parent();
                  firstImg.remove();
                  // Remove empty parent containers recursively
                  let currentParent = parent;
                  while (currentParent.length > 0) {
                    const tagName = currentParent
                      .get(0)
                      ?.tagName?.toLowerCase();
                    if (tagName === "body" || tagName === "html") {
                      break;
                    }
                    const text = currentParent.text().trim();
                    const hasChildren = currentParent.children().length > 0;
                    if (!text && !hasChildren) {
                      const nextParent = currentParent.parent();
                      currentParent.remove();
                      currentParent = nextParent;
                    } else {
                      break;
                    }
                  }
                  logger.debug(
                    "Removed first image from content (using YouTube video instead)",
                  );
                }
              }

              // Remove the original link/image element if it was in the content
              if (firstElement && firstElement.length > 0) {
                const parent = firstElement.parent();
                firstElement.remove();
                // Remove empty parent containers recursively
                let currentParent = parent;
                while (currentParent.length > 0) {
                  const tagName = currentParent.get(0)?.tagName?.toLowerCase();
                  if (tagName === "body" || tagName === "html") {
                    break;
                  }
                  const text = currentParent.text().trim();
                  const hasChildren = currentParent.children().length > 0;
                  if (!text && !hasChildren) {
                    const nextParent = currentParent.parent();
                    currentParent.remove();
                    currentParent = nextParent;
                  } else {
                    break;
                  }
                }
                logger.debug("Removed original link/image from content");
              }
            } else {
              logger.debug(
                { url: firstUrl },
                "Failed to create header element from URL",
              );
            }
          }
        } else {
          logger.debug(
            { firstUrl },
            "Skipping header element creation for invalid URL",
          );
        }
      }
    }

    // Check if footer already exists before removing anything
    const hasExistingFooter = $body("footer").length > 0;

    // Extract existing headers (they should be preserved as-is)
    // This is important for aggregators like Tagesschau that add custom media headers
    const existingHeaders: string[] = [];
    $body("header").each((_, el) => {
      const $header = $body(el);
      existingHeaders.push($header.toString());
    });

    // Extract comment sections (they should be preserved as-is)
    const commentSections: string[] = [];
    $body("section").each((_, el) => {
      const $section = $body(el);
      // Check if this section contains comments (has "Comments" in heading or text)
      const sectionText = $section.text().toLowerCase();
      const sectionHtml = $section.html() || "";
      // Preserve sections that contain "Comments" (case-insensitive) or have comment-like structure
      if (
        sectionText.includes("comment") ||
        sectionHtml.match(/<h[1-6][^>]*>.*[Cc]omment/i)
      ) {
        commentSections.push($section.toString());
      }
    });

    // Remove header, footer, and comment sections from body content
    $body("header").remove();
    $body("footer").remove();
    $body("section").remove();

    const bodyHtml = $body.html() || "";

    // Build the final structure
    // Prefer existing headers (from aggregators like Tagesschau) over newly generated ones
    const existingHeaderHtml = existingHeaders.join("");
    const newHeaderHtml =
      contentParts.find((part) => part.includes("<header>")) || "";
    // Use existing header if present, otherwise use newly generated header
    const headerHtml = existingHeaderHtml || newHeaderHtml;

    // Wrap main content in section tag (unless it's empty)
    const mainContentSection = bodyHtml.trim()
      ? `<section>${bodyHtml}</section>`
      : "";

    // Preserve comment sections
    const commentSectionsHtml = commentSections.join("");

    const footerHtml =
      addSourceFooter && !hasExistingFooter
        ? `<footer><a href="${article.url}" style="float: right;">Source</a></footer>`
        : "";

    // Wrap everything in <article> tag
    const articleContent = `<article>${headerHtml}${mainContentSection}${commentSectionsHtml}${footerHtml}</article>`;

    return articleContent;
  } catch (error) {
    logger.error({ error }, "Error standardizing content format");
    // Fallback: wrap in article tag with footer if enabled
    const $ = cheerio.load(content);
    const hasExistingFooter = $("footer").length > 0;

    let fallbackContent = content;
    if (addSourceFooter && !hasExistingFooter) {
      fallbackContent = `<article>${content}<footer><a href="${article.url}" style="float: right;">Source</a></footer></article>`;
    } else if (!$("article").length) {
      fallbackContent = `<article>${content}</article>`;
    }

    return fallbackContent;
  }
}

/**
 * Process and sanitize HTML content.
 */
export async function processContent(
  html: string,
  article: RawArticle,
  generateTitleImage: boolean = true,
  addSourceFooter: boolean = true,
  headerImageUrl?: string,
): Promise<string> {
  // Standardize format (add header image, source link)
  return await standardizeContentFormat(
    html,
    article,
    article.url,
    generateTitleImage,
    addSourceFooter,
    headerImageUrl,
  );
}
